lib/events/dynamoevents/dynamoevents.go

```diff
@@ -500,74 +643,87 @@ func (l *Log) SearchEvents(fromUTC, toUTC time.Time, filter string, limit int) (
 	doFilter := len(eventFilter) > 0
 
 	var values []events.EventFields
-	query := "EventNamespace = :eventNamespace AND CreatedAt BETWEEN :start and :end"
-	attributes := map[string]interface{}{
-		":eventNamespace": defaults.Namespace,
-		":start":          fromUTC.Unix(),
-		":end":            toUTC.Unix(),
-	}
-	attributeValues, err := dynamodbattribute.MarshalMap(attributes)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
+	dates := daysBetween(fromUTC, toUTC)
+	query := "CreatedAtDate = :date AND CreatedAt BETWEEN :start and :end"
 
-	var lastEvaluatedKey map[string]*dynamodb.AttributeValue
 	var total int
 
-	// Because the maximum size of the dynamo db response size is 900K according to documentation,
-	// we arbitrary limit the total size to 100MB to prevent runaway loops.
-	for pageCount := 0; pageCount < 100; pageCount++ {
-		input := dynamodb.QueryInput{
-			KeyConditionExpression:    aws.String(query),
-			TableName:                 aws.String(l.Tablename),
-			ExpressionAttributeValues: attributeValues,
-			IndexName:                 aws.String(indexTimeSearch),
-			ExclusiveStartKey:         lastEvaluatedKey,
+dateLoop:
+	for _, date := range dates {
```

Comment:
This does O(N) queries. Is there any way we can do a single query over only `CreatedAt` without  `CreatedAtDate`?

Reply:
Unfortunately the index partitioning disallows this due to how DynamoDB works internally, this is what we do at the moment on master but due to partitioning constraints this means all data ends up in one hardcoded partition which is the precise thing RFD 24 works around since partitions have data limits.

Reply:
Had another look through the API and this seems to be impossible with current API limitations. `BatchGetRequest` looked interesting at first but it turns out it only supports point lookups for range keys which won't work for our case.